numpy/core/src/multiarray/compiled_base.c

```diff
@@ -603,9 +609,15 @@ arr_interp(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwdict)
                 dres[i] = dy[j];
             }
             else {
-                const npy_double slope = (slopes != NULL) ? slopes[j] :
-                                         (dy[j+1] - dy[j]) / (dx[j+1] - dx[j]);
+                const npy_double slope =
+                        (slopes != NULL) ? slopes[j] :
+                        _subtract_not_nan(dy[j+1], dy[j]) / (dx[j+1] - dx[j]);
+
+                /* If we get nan in one direction, try the other */
                 dres[i] = slope*(x_val - dx[j]) + dy[j];
+                if (NPY_UNLIKELY(npy_isnan(dres[i]))) {
+                    dres[i] = slope*(x_val - dx[j+1]) + dy[j+1];
+                }
```

Comment:
Perhaps instead of `_subtract_not_nan`, I should add one more clause here:
```
if (NPY_UNLIKELY(npy_isnan(dres[i])) && dy[j+1] == d[y]) {
    dres[i] = dy[j];
}
```

Reply:
Done, this helps significantly, and catches a case I missed too.